### DISCUSSION TITLE
Can I import and use a pre-trained model with TensorFlow with do-mpc?

### DISCUSSION THREAD

**Discussion by nacereddinemoumai** [do-mpc/do-mpc#458] [General]
I am currently implementing a Model Predictive Control (MPC) system to control the temperature inside a room. To achieve this, I have used an LSTM neural network with TensorFlow to predict the output over a prediction horizon. My question is: can we use this pre-trained model with do-mpc to formulate a data-driven MPC problem?

**Comment by 4flixt** [comment]
Hi @nacereddinemoumai ,

thank you for your question! Unfortunately, this is not straight-forward. While do-mpc supports the automatic conversion of neural network models (saved in the ONNX format), the tool which is used for setting up the optimal control problem problem (CasADi) does not support higher-order tensors. Since the LSTM model uses higher order tensors, the conversion is not supported. With a lot of manual work (and if the inputs of your LSTM are vectors) you can formulate the LSTM (with your trained weights) in CasADi yourself and then use do-mpc.
Obviously, this is not ideal of course. 

Sorry to disappoint and hope you find a solution.

Best,
4flixt

**Reply by nacereddinemoumai** [reply]
Hi @4flixt , Thank you for taking the time to answer my question. Otherwise, I read that CasADi only supports PyTorch models. In that case, I need to have a PyTorch model, right?

**Comment by 4flixt** [comment]
CasADi by itself has nothing to do with Pytorch or Tensorflow. We bundle do-mpc with an ONNX conversion tool. ONNX models can be exported from both, Pytorch and Tensorflow.
